File: java/src/main/java/com/example/design_pattern/Composite.java
package com.example.design_pattern;

import java.util.*;

public class Composite {

  private static class ArrayBasedAttribute {

    private static class Creature {

      public enum Abilites {
        STR,
        AGL,
        INTL,
        COUNT;
      }

      private EnumMap<Abilites, Integer> abilites;

      Creature(int strength, int agility, int intelligence) {
        abilites = new EnumMap<>(Abilites.class);
        abilites.put(Abilites.STR, strength);
        abilites.put(Abilites.AGL, agility);
        abilites.put(Abilites.INTL, intelligence);
      }

      public int sum() {
        return abilites.values().stream().mapToInt(i -> i).sum();
      }

      public double average() {
        return sum() / (abilites.size());
      }

      public int max() {
        return abilites.values().stream().mapToInt(i -> i).max().getAsInt();
      }
    }

    public static void main() {
      Creature creature = new Creature(1, 2, 3);
      System.out.println(creature.max());
    }
  }

  private static class Composite_ {

    private static interface GraphicObject {
      void draw();
    }

    private static class Circle implements GraphicObject {
      @Override
      public void draw() {
        System.out.println("Circle");
      }
    }

    private static class Group implements GraphicObject {
      private String name;
      private List<GraphicObject> objects;

      Group(String name) {
        this.name = name;
        objects = new LinkedList<>();
      }

      public void add(GraphicObject object) {
        objects.add(object);
      }

      @Override
      public void draw() {
        System.out.printf("group <%s> contains : ", name);
        for (GraphicObject object : objects) {
          object.draw();
        }
      }
    }

    public static void main() {
      Group root = new Group("root");
      Group subGroup = new Group("sub");
      Circle c1 = new Circle();
      Circle c2 = new Circle();

      subGroup.add(c2);
      root.add(c1);
      root.add(subGroup);
      root.draw();
    }
  }

  private static class TemplateBasedComposite {

    private abstract static class NeuronBase implements Iterable<Neuron> {
      public <T extends NeuronBase> void connectTo(T other) {
        for (Neuron in : this) {
          for (Neuron out : other) {
            out.connect(in);
          }
        }
      }

      public abstract Iterator<Neuron> iterator();
    }

    private static class Neuron extends NeuronBase {
      private static int number = 1;
      private List<Neuron> in;
      private List<Neuron> out;
      private int id;

      private class SelfIterator implements Iterator<Neuron> {
        private boolean hasNext = false;

        @Override
        public boolean hasNext() {
          return hasNext;
        }

        @Override
        public Neuron next() {
          hasNext = false;
          return Neuron.this;
        }

        @Override
        public void remove() {}
      }

      Neuron() {
        in = new LinkedList<>();
        out = new LinkedList<>();
        id = number++;
      }

      public void connect(Neuron other) {
        other.in.add(this);
        out.add(other);
      }

      @Override
      public Iterator<Neuron> iterator() {
        return new SelfIterator();
      }
    }

    private static class NeuronLayer extends NeuronBase {
      private List<Neuron> layer;

      public NeuronLayer(int count) {
        this.layer = new LinkedList<>();
        while (count > 0) {
          layer.add(new Neuron());
          count--;
        }
      }

      @Override
      public Iterator<Neuron> iterator() {
        return layer.iterator();
      }
    }

    public static void main() {
      Neuron n1 = new Neuron();
      Neuron n2 = new Neuron();
      NeuronLayer layer = new NeuronLayer(5);
      n1.connectTo(layer);
      layer.connectTo(n2);
    }
  }

  public static void main(String... args) {
    ArrayBasedAttribute.main();
    Composite_.main();
    TemplateBasedComposite.main();
  }
}

File: c++/design_pattern/Composite.cpp
#include <array>
#include <iostream>
#include <numeric>

using namespace std;

namespace ArrayBasedAttribute {

class BadCaseCreature {
  int strength, agility, intelligence;

 public:
  int sum() const { return strength + agility + intelligence; }

  double average() const { return sum() / 3; }

  int max() const { return ::max(::max(strength, agility), intelligence); }
};

class GoodCaseCreature {
  enum Abilites { str, agl, intl, count };

  array<int, count> abillites;

 public:
  GoodCaseCreature(int strength, int agility, int intelligence) {
    abillites[str] = strength;
    abillites[agl] = agility;
    abillites[intl] = intelligence;
  }

  int sum() const { return accumulate(begin(abillites), end(abillites), 0); }

  double average() const { return sum() / (double)count; }

  int max() const { return *max_element(begin(abillites), end(abillites)); }
};

void main() {
  GoodCaseCreature creature1{1, 0, 5};
  cout << creature1.max() << endl;
}
};  // namespace ArrayBasedAttribute

namespace Composite {

struct GraphicObject {
  virtual void draw() = 0;
};

struct Circle : GraphicObject {
  void draw() override { cout << "Circle" << endl; }
};

struct Group : GraphicObject {
  vector<GraphicObject*> objects;
  string name;

  Group(const string& name) : name(name) {}

  void draw() override {
    cout << "Group <" << name << "> contians : ";
    for (auto& o : objects) {
      o->draw();
    }
  }
};

void main() {
  Group root("root");
  Circle c1, c2;
  root.objects.push_back(&c1);

  Group subgroup("sub");
  subgroup.objects.push_back(&c2);
  root.objects.push_back(&subgroup);

  root.draw();
}

};  // namespace Composite

namespace TamplateBasedComposite {

// 1. 개별객체와 컬렉션 객체 모두에 동일한 인터페이스를 사용가능하도록 template
// method

template <typename Self>
struct NeuronBase;
struct Neuron;

template <typename Self>
struct NeuronBase {
  template <typename T>
  void connect_to(T& ohter);
};

struct Neuron : NeuronBase<Neuron> {
  vector<Neuron*> in, out;
  unsigned int id;

  Neuron() {
    static int id = 1;
    this->id = id++;
  }

  Neuron* begin();
  Neuron* end();
};

struct NeuronLayer : vector<Neuron>, NeuronBase<NeuronLayer> {
  NeuronLayer(int count) {
    while (count-- > 0) emplace_back(Neuron{});
  }
};

template <typename Self>
template <typename T>
void NeuronBase<Self>::connect_to(T& other) {
  for (Neuron& from : *static_cast<Self*>(this)) {
    for (Neuron& to : other) {
      from.out.push_back(&to);
      to.in.push_back(&from);
    }
  }
}

Neuron* Neuron::begin() { return this; }
Neuron* Neuron::end() { return this + 1; }

void main() {
  Neuron n1, n2;
  NeuronLayer layer1(0), layer2(0);
  n1.connect_to(n2);
  n1.connect_to(layer1);
  layer1.connect_to(n1);
  layer1.connect_to(layer2);
}

};  // namespace TamplateBasedComposite

int main(void) {
  ArrayBasedAttribute::main();
  Composite::main();
  TamplateBasedComposite::main();

  return 0;
}